migrating-to-cd4pe.md

```diff
@@ -0,0 +1,85 @@
+
+* `production`
+* `staging`
+* `development`
+
+Usually changes are worked on in "feature branches" which contain a single feature and are reviewed though a pull request into one of the long lived branches. The destination of these pull requests depends on when the code integration is supposed to happen, however the exact details of the workflow does not matter.
+
+### Trunk Workflows
+
+Some workflows use a central branch such as `development` which features branches are taken from and merged back into. Occasionally releases are taken from this branch and merged into another long-lived branch such as `production`. Sometimes this also is combined with a "release branch" system where the release branch is tested before being merged. It is also possible to combine this with many other long-lived branches that the release would pass through before making it to production.
+
+## Best Practice Details
+
+Let's assume that we are beginning from a state where we have three branches with the following refs:
+
+| Branch | SHA | Logical Order |
+|--------|-----|---------------|
+| production | `34211f1` | 3rd |
+| staging | `f0f25e0` | 2nd |
+| development | `593a1e3` | 1st |
+
+In order to ensure that we can have CD4PE deploy new bookkeeping branches with the same revisions as the existing branches (and therefore not cause any unnecessary changes) we need to create a `master` branch which contains all of the commits in the repository.
+
+### Perform the Merge-Back
+
+Before creating the `master` branch, perform a merge-back. To do this we need to merge the long lived branches back onto each other in reverse order. In the above environment this would mean:
+
+* Create a pull request from `production` to `staging`
+* Review the changes and merge the PR
+* Create a pull request from `staging` to `development`
+
+In each PR the only changes that should appear in the diff are ones that were not present in the environment above but **not** in the environment below. As long is the logical order was followed, this should never happen. There will however likely be a large number of merge commits. If there are changes, these should be reviewed and merged.
+
+Once all branches have been merged back to the 1st branch (`development`), take a note of the new refs:
+
+| Branch | SHA | Logical Order |
+|--------|-----|---------------|
+| production | `34211f1` | 3rd |
+| staging | `5380862` | 2nd |
+| development | `52c74cb` | 1st |
+
+### Create the `master` branch
+
+Once the merge-back is complete, create a branch named `master` from the 1st logical branch (`development`). This should contain all of the commits that the current long-lived branches are using. This can be checked using the following command which will return the name of the master branch if it contains the commit.
+
+```shell
+$ git branch --contains e0aaea8
+  development
+* master
+```
+
+### Create the bookkeeping branches
```

Comment:
I get lost here, I'm not sure why all the bookkeeping branches are needed. Is there a place in CD4PE docs that talks about it?

Reply:
I don't think that anything really covers how CD4PE uses branches, the closest thing is here: https://puppet.com/docs/continuous-delivery/3.x/deployment_policies.html and it doesn't use the term "bookkeeping". We should have a word for them because while they are a branch, they are used more like a moveable tag and nobody would ever interact with them manually.

@reidmv this is me finally writing down what we discussed in PDX. Any comments?

Reply:
I do agree that the whole thing kind of falls apart towards the end...